### ISSUE TITLE
Fix_imbalance is creating large volume of data and setup is taking too much time

### ISSUE THREAD

**Issue by sathishkumar999** [pycaret/pycaret#3192]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

Hi pycaret team,
I tested the pycaret 3.0 -pre version. I'm working on ticket priority classfication. it has around 320k records. and 3 classes in target class. 
it's basically text data and we have to classify the priority whether it's high, moderate and low priority.
here our input is short_desription and target is priority. and data has imbalanced data. out of 320k records, 280 K Records are medium and remaining records are high and low. in setup I enabled fix_imbalance = True.
after running setup, it has created 1 billion records(10 million). it shoud create around 810k records. because smote model creates equal number of records for majority class. such as medium = 270k, high= 270k and low =270k. instead of that it's created 1 billion records and it took 48 minutes to run the setup and for training models it was taking more than 4 hours, it was still running. so i interrupted kernel. it's stange fix_imbalnce is created huge number of records

### Reproducible Example

```python
#df1['priority'].value_counts()
Moderate    269989
Urgent       50419
Critical      2590
#df1.shape
(322998, 2)
#from pycaret.classification import *
%%time
clf = setup(data=df1,target = 'priority', fix_imbalance = True,train_size = 0.8)
```

### Expected Behavior

transformed data shape should have 810K records around
High = 270k
Medium = 270k
Low=270k
![Screenshot (1652)](https://user-images.githubusercontent.com/40235632/208818110-d667344a-eebb-48ef-ad46-941f876979ae.png)
![Screenshot (1653)](https://user-images.githubusercontent.com/40235632/208818117-0fd6a243-89aa-4f7b-baba-bce1a33a79b0.png)
![Screenshot (1654)](https://user-images.githubusercontent.com/40235632/208818119-6ebcae9f-d629-4b32-8e48-8171a81ebbcf.png)

### Actual Results

```python-traceback
Description	Value
0	Session id	6618
1	Target	Priority
2	Target type	Multiclass
3	Target mapping	High: 0, Low: 1, Medium: 2
4	Original data shape	(10435094, 2)
5	Transformed data shape	(23021259, 2)
6	Transformed train set shape	(20934240, 2)
7	Transformed test set shape	(2087019, 2)
8	Categorical features	1
9	Preprocess	True
10	Imputation type	simple
11	Numeric imputation	mean
12	Categorical imputation	constant
13	Maximum one-hot encoding	5
14	Encoding method	None
15	Low variance threshold	0
16	Fix imbalance	True
17	Fix imbalance method	SMOTE
18	Fold Generator	StratifiedKFold
19	Fold Number	10
20	CPU Jobs	-1
21	Use GPU	False
22	Log Experiment	False
23	Experiment Name	clf-default-name
24	USI	d02c
```

### Installed Versions

pycaret.__version__

'3.0.0.rc4'

**Comment by sathishkumar999** [comment]
@ngupta23 
Here is the code for running the data and data set
[Uploading kaggle_ticket_data.csv…]()

import pandas as pd
df1 = pf.read_csv("kaggle_ticket_data.csv")
df.head()
df1.columns
df1['priority'].value_counts()
Moderate    269989
Urgent       50419
Critical      2590
df1.shape
(322998, 2)
from pycaret.classification import *
%%time
clf = setup(data=df1,target = 'priority', fix_imbalance = True,train_size = 0.8)
```


### Expected Behavior

transformed data should have equal number of records for fix_imbalace = True. And it should be around 810k records
Moderate = 269989
Urgent = 269989
Critical=269989

But transformed data has more than 10 million records
### Actual Results

```python-traceback
Description	Value
0	Session id	6618
1	Target	Priority
2	Target type	Multiclass
3	Target mapping	High: 0, Low: 1, Medium: 2
4	Original data shape	(10435094, 2)
5	Transformed data shape	(23021259, 2)
6	Transformed train set shape	(20934240, 2)
7	Transformed test set shape	(2087019, 2)
8	Categorical features	1
9	Preprocess	True
10	Imputation type	simple
11	Numeric imputation	mean
12	Categorical imputation	constant
13	Maximum one-hot encoding	5
14	Encoding method	None
15	Low variance threshold	0
16	Fix imbalance	True
17	Fix imbalance method	SMOTE
18	Fold Generator	StratifiedKFold
19	Fold Number	10
20	CPU Jobs	-1
21	Use GPU	False
22	Log Experiment	False
23	Experiment Name	clf-default-name
24	USI	d02c
```


### Installed Versions

pycaret.__version__

'3.0.0.rc4'

**Comment by sathishkumar999** [comment]
@ngupta23 , @moezali1 
Please look into abovd issue. It's been 3 months. Hope you will fix the issue soon

**Comment by Yard1** [comment]
Isn't it expected with SMOTE? It is a very performance and memory heavy algorithm. Perhaps you should use undersampling instead.

**Comment by sathishkumar999** [comment]
@Yard1 
Thanks for you're response. I believe it used smote model by default for Imbalance=True.
But for smote model it should create equal number of values for all categorical values.
Example:
Here majority class is,
Moderate:269989
And other classes are high and low classes.
So when we use smote it should create equal number of duplicate values for both high and low classes.
Such as high=269989, medium=269989 and low=269989.
I think something wrong with smote model

If it's possible please create pycaret custom parameters for Imbalance. For example if imbalance=true. It will find which class has majority values and it will create equal number of classes which has minor values. Or else the user can mention the percentage of duplicate values that he wants create for minor values.

There is an issue with under sampling, because in undersampling we will lose original data.

I suspect here the issue with text data. When use numerical data and categorical data, I didn't see issue like this. But when I'm using text data,. It's creating issue. 
Actually I used the tfidf and count vectorizer pipeline for this text data using sklearn. And it took 30 minutes for training data.

But here in pycaret, just for creating setup imbalance it's taking 50 minutes.
Can you please provide parallel processor code like dask or fuge for training large data and also use tfidf,count vectorizer for text data. Instead of label or one hot encoding. And create pycaret custom imbalance parameter. Such as Imbalance= majority or Imbalance=minority

Here Imbalance=majority(oversampling) will create duplicate values for minority class which is equal to majority class.

Imbalance=minority(undersampling) will create duplicate values for majority class which is equal to minority class

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
